// @prisma/language-server/dist/src/__test__/documentSymbol.test.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
const vscode_languageserver_1 = require("vscode-languageserver");
const assert = __importStar(require("assert"));
const MessageHandler_1 = require("../MessageHandler");
const helper_1 = require("./helper");
function assertSymbols(fixturePath, expected) {
    const textDocument = (0, helper_1.getTextDocument)(fixturePath);
    const actual = (0, MessageHandler_1.handleDocumentSymbol)({ textDocument }, textDocument);
    assert.deepStrictEqual(actual, expected);
}
suite('DocumentSymbol', () => {
    test('hover_postgresql.prisma', () => {
        assertSymbols('./hover_postgresql.prisma', [
            {
                kind: vscode_languageserver_1.SymbolKind.Struct,
                name: 'db',
                range: {
                    end: {
                        character: 1,
                        line: 3,
                    },
                    start: {
                        character: 0,
                        line: 0,
                    },
                },
                selectionRange: {
                    end: {
                        character: 13,
                        line: 0,
                    },
                    start: {
                        character: 11,
                        line: 0,
                    },
                },
            },
            {
                kind: vscode_languageserver_1.SymbolKind.Function,
                name: 'client',
                range: {
                    end: {
                        character: 1,
                        line: 7,
                    },
                    start: {
                        character: 0,
                        line: 5,
                    },
                },
                selectionRange: {
                    end: {
                        character: 16,
                        line: 5,
                    },
                    start: {
                        character: 10,
                        line: 5,
                    },
                },
            },
            {
                kind: vscode_languageserver_1.SymbolKind.Class,
                name: 'Post',
                range: {
                    end: {
                        character: 1,
                        line: 16,
                    },
                    start: {
                        character: 0,
                        line: 10,
                    },
                },
                selectionRange: {
                    end: {
                        character: 10,
                        line: 10,
                    },
                    start: {
                        character: 6,
                        line: 10,
                    },
                },
            },
            {
                kind: vscode_languageserver_1.SymbolKind.Class,
                name: 'User',
                range: {
                    end: {
                        character: 1,
                        line: 26,
                    },
                    start: {
                        character: 0,
                        line: 19,
                    },
                },
                selectionRange: {
                    end: {
                        character: 10,
                        line: 19,
                    },
                    start: {
                        character: 6,
                        line: 19,
                    },
                },
            },
            {
                kind: vscode_languageserver_1.SymbolKind.Enum,
                name: 'UserName',
                range: {
                    end: {
                        character: 1,
                        line: 32,
                    },
                    start: {
                        character: 0,
                        line: 29,
                    },
                },
                selectionRange: {
                    end: {
                        character: 13,
                        line: 29,
                    },
                    start: {
                        character: 5,
                        line: 29,
                    },
                },
            },
            {
                kind: vscode_languageserver_1.SymbolKind.Enum,
                name: 'Test',
                range: {
                    end: {
                        character: 1,
                        line: 38,
                    },
                    start: {
                        character: 0,
                        line: 35,
                    },
                },
                selectionRange: {
                    end: {
                        character: 9,
                        line: 35,
                    },
                    start: {
                        character: 5,
                        line: 35,
                    },
                },
            },
        ]);
    });
    test('hover_mongodb.prisma', () => {
        assertSymbols('./hover_mongodb.prisma', [
            {
                kind: vscode_languageserver_1.SymbolKind.Function,
                name: 'client',
                range: {
                    end: {
                        character: 1,
                        line: 2,
                    },
                    start: {
                        character: 0,
                        line: 0,
                    },
                },
                selectionRange: {
                    end: {
                        character: 16,
                        line: 0,
                    },
                    start: {
                        character: 10,
                        line: 0,
                    },
                },
            },
            {
                kind: vscode_languageserver_1.SymbolKind.Struct,
                name: 'db',
                range: {
                    end: {
                        character: 1,
                        line: 7,
                    },
                    start: {
                        character: 0,
                        line: 4,
                    },
                },
                selectionRange: {
                    end: {
                        character: 13,
                        line: 4,
                    },
                    start: {
                        character: 11,
                        line: 4,
                    },
                },
            },
            {
                kind: vscode_languageserver_1.SymbolKind.Class,
                name: 'Post',
                range: {
                    end: {
                        character: 1,
                        line: 15,
                    },
                    start: {
                        character: 0,
                        line: 9,
                    },
                },
                selectionRange: {
                    end: {
                        character: 10,
                        line: 9,
                    },
                    start: {
                        character: 6,
                        line: 9,
                    },
                },
            },
            {
                kind: vscode_languageserver_1.SymbolKind.Class,
                name: 'User',
                range: {
                    end: {
                        character: 1,
                        line: 24,
                    },
                    start: {
                        character: 0,
                        line: 17,
                    },
                },
                selectionRange: {
                    end: {
                        character: 10,
                        line: 17,
                    },
                    start: {
                        character: 6,
                        line: 17,
                    },
                },
            },
            {
                kind: vscode_languageserver_1.SymbolKind.Class,
                name: 'EmbedHolder',
                range: {
                    end: {
                        character: 1,
                        line: 35,
                    },
                    start: {
                        character: 0,
                        line: 26,
                    },
                },
                selectionRange: {
                    end: {
                        character: 17,
                        line: 26,
                    },
                    start: {
                        character: 6,
                        line: 26,
                    },
                },
            },
            {
                kind: vscode_languageserver_1.SymbolKind.Interface,
                name: 'Embed',
                range: {
                    end: {
                        character: 1,
                        line: 44,
                    },
                    start: {
                        character: 0,
                        line: 37,
                    },
                },
                selectionRange: {
                    end: {
                        character: 10,
                        line: 37,
                    },
                    start: {
                        character: 5,
                        line: 37,
                    },
                },
            },
            {
                kind: vscode_languageserver_1.SymbolKind.Interface,
                name: 'EmbedEmbed',
                range: {
                    end: {
                        character: 1,
                        line: 49,
                    },
                    start: {
                        character: 0,
                        line: 46,
                    },
                },
                selectionRange: {
                    end: {
                        character: 15,
                        line: 46,
                    },
                    start: {
                        character: 5,
                        line: 46,
                    },
                },
            },
        ]);
    });
});
